lib/js/src/manager/screen/BaseVoiceCommandManager.js

```diff
@@ -0,0 +1,219 @@
+        this._isUpdateRunning = false; // prevent _update from being invoked more than once at a time
+        this._addListeners();
+    }
+
+    /**
+     * @return {Promise}
+    */
+    async start () {
+        this._transitionToState(BaseSubManager.READY);
+        super.start();
+    }
+
+    dispose () {
+        this._currentHmiLevel = HMILevel.HMI_NONE;
+        this._taskQueue = [];
+        this._voiceCommands = [];
+        this._lastVoiceCommandId = this._voiceCommandIdMin;
+        this._isReady = false;
+        // remove listeners
+        this._lifecycleManager.removeRpcListener(FunctionID.ON_HMI_STATUS, this._hmiListener);
+        this._lifecycleManager.removeRpcListener(FunctionID.ON_COMMAND, this._commandListener);
+        this._hmiListener = null;
+        this._commandListener = null;
+        this._isUpdateRunning = false;
+
+        super.dispose();
+    }
+
+    /**
+     * @param {VoiceCommand[]} voiceCommands
+     * @return {Promise} - returns after old commands are deleted and new ones are added
+    */
+    async setVoiceCommands (voiceCommands) {
+        // we actually need voice commands to set.
+        if (!Array.isArray(voiceCommands)) {
+            return;
+        }
+        // add the commands to a queue to be processed later
+        this._taskQueue.push(voiceCommands);
+
+        if (!this._isUpdateRunning) {
+            this._isUpdateRunning = true;
+            await this._update();
+            this._isUpdateRunning = false;
+        }
+    }
+
+    /**
+     * Looks through the task queue to process incoming voice commands. Recursive
+     * @return {Promise} - returns after old commands are deleted and new ones are added
+    */
+    async _update () {
+        // we actually need voice commands to set.
+        if (this._taskQueue.length === 0) {
+            return;
+        }
+
+        if (!this._isReady) { // not in the right HMI level yet
+            return;
+        }
+
+        // remove all but the last element in the task queue for processing
+        // there is only going to be one voice command set in the end
+        // any other than the last element will end up being deleted anyway
+        let voiceCommands = [];
+        while (this._taskQueue.length > 0) {
+            voiceCommands = this._taskQueue.shift();
+        }
+
+        this._lastVoiceCommandId = this._voiceCommandIdMin;
+        for (const voiceCommand of voiceCommands) {
+            this._lastVoiceCommandId++;
+            voiceCommand.setCommandId(this._lastVoiceCommandId);
+        }
+
+        await this._sendDeleteVoiceCommands(this._voiceCommands);
+        // old voice commands are now deleted
+        // now add the new ones
+        await this._sendAddVoiceCommands(voiceCommands);
+        this._voiceCommands = voiceCommands;
+        // run this method again for potential future tasks
+        return this._update();
+    }
+
+    /**
+     * @return {VoiceCommand[]}
```

Comment:
Please ensure all methods have descriptions (applies to multiple methods in this file)